File: Roadmap/04 - CADENAS DE CARACTERES/c/N0HagoNada.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

#define MAX_LEN 100

// Función para comprobar si una palabra es un palíndromo
int esPalindromo(char *palabra)
{
    int longitud = strlen(palabra);
    for (int i = 0; i < longitud / 2; i++)
    {
        if (tolower(palabra[i]) != tolower(palabra[longitud - i - 1]))
        {
            return 0; // No es palíndromo
        }
    }
    return 1; // Es palíndromo
}

// Función para comprobar si dos palabras son anagramas
int sonAnagramas(char *palabra1, char *palabra2)
{
    int count[256] = {0};
    int i;

    for (i = 0; palabra1[i] && palabra2[i]; i++)
    {
        count[(int)tolower(palabra1[i])]++;
        count[(int)tolower(palabra2[i])]--;
    }

    if (palabra1[i] || palabra2[i])
    { // Si las palabras son de diferente longitud
        return 0;
    }

    for (i = 0; i < 256; i++)
    {
        if (count[i])
        {
            return 0; // No son anagramas
        }
    }
    return 1; // Son anagramas
}

// Función para comprobar si una palabra es un isograma
int esIsograma(char *palabra)
{
    int count[256] = {0};

    for (int i = 0; palabra[i]; i++)
    {
        if (count[(int)tolower(palabra[i])])
        {
            return 0; // No es isograma
        }
        count[(int)tolower(palabra[i])]++;
    }
    return 1; // Es isograma
}
int main()
{
    // Definición de la cadena original
    char str[] = "Hola mundo";
    char copia[20]; // Para operaciones que necesitan una copia

    // Acceso a caracteres específicos
    printf("Primer caracter: %c\n", str[0]);

    // Subcadenas
    strncpy(copia, str + 5, 5); // Copia "mundo" en copia
    copia[5] = '\0';            // Añade manualmente el carácter nulo
    printf("Subcadena: %s\n", copia);

    // Longitud
    printf("Longitud: %lu\n", strlen(str));

    // Concatenación
    char saludo[20] = "Hola";
    strcat(saludo, ", mundo");
    printf("Concatenación: %s\n", saludo);

    // Conversión a mayúsculas
    for (int i = 0; str[i]; i++)
    {
        str[i] = toupper(str[i]);
    }
    printf("A mayúsculas: %s\n", str);

    // Conversión a minúsculas
    for (int i = 0; str[i]; i++)
    {
        str[i] = tolower(str[i]);
    }
    printf("A minúsculas: %s\n", str);

    // Recorrido
    printf("Recorrido: ");
    for (int i = 0; str[i]; i++)
    {
        printf("%c", str[i]);
    }
    printf("\n");

    // División (Tokenización)
    char str2[] = "Hola mundo, prueba"; // Nueva cadena para evitar modificar la original en la tokenización
    printf("División:\n");
    char *token = strtok(str2, " ,");
    while (token != NULL)
    {
        printf("%s\n", token);
        token = strtok(NULL, " ,");
    }

    // Verificación (Buscar Subcadena)
    strcpy(copia, "Hola mundo"); // Restablecer copia a un valor conocido
    if (strstr(copia, "mundo") != NULL)
    {
        printf("Subcadena encontrada.\n");
    }
    else
    {
        printf("Subcadena no encontrada.\n");
    }

    char palabra1[MAX_LEN], palabra2[MAX_LEN];

    printf("Introduce la primera palabra: ");
    scanf("%99s", palabra1);
    printf("Introduce la segunda palabra: ");
    scanf("%99s", palabra2);

    // Comprobar si las palabras son palíndromos
    printf("'%s' es un palíndromo? %s\n", palabra1, esPalindromo(palabra1) ? "Sí" : "No");
    printf("'%s' es un palíndromo? %s\n", palabra2, esPalindromo(palabra2) ? "Sí" : "No");

    // Comprobar si las palabras son anagramas
    printf("'%s' y '%s' son anagramas? %s\n", palabra1, palabra2, sonAnagramas(palabra1, palabra2) ? "Sí" : "No");

    // Comprobar si las palabras son isogramas
    printf("'%s' es un isograma? %s\n", palabra1, esIsograma(palabra1) ? "Sí" : "No");
    printf("'%s' es un isograma? %s\n", palabra2, esIsograma(palabra2) ? "Sí" : "No");

    return 0;
}

File: Roadmap/04 - CADENAS DE CARACTERES/go/N0HagoNada.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Println("Introduce una palabra (o escribe 'exit' para salir):")
		scanner.Scan()
		palabra := strings.TrimSpace(scanner.Text())

		if palabra == "exit" {
			break
		}

		if esPalindromo(palabra) {
			fmt.Println(palabra, "es un palíndromo.")
		} else {
			fmt.Println(palabra, "no es un palíndromo.")
		}

		if esIsograma(palabra) {
			fmt.Println(palabra, "es un isograma.")
		} else {
			fmt.Println(palabra, "no es un isograma.")
		}

		fmt.Println("Introduce otra palabra para comparar si son anagramas (o escribe 'exit' para salir):")
		scanner.Scan()
		otraPalabra := strings.TrimSpace(scanner.Text())

		if otraPalabra == "exit" {
			break
		}

		if esAnagrama(palabra, otraPalabra) {
			fmt.Println(palabra, "y", otraPalabra, "son anagramas.")
		} else {
			fmt.Println(palabra, "y", otraPalabra, "no son anagramas.")
		}
	}
	operacioneStrings()
	TestStringIterationWithRange()
	TestStringEqualsCaseInsensitive()
}

func operacioneStrings() {
	// Imprimir cadena
	fmt.Println("Let's print out this string.")
	// Concatenar
	fmt.Println("Sammy" + "Shark")
	// Aplicar mayúsculas y minúsculas
	ss := "Sammy Shark"
	fmt.Println(strings.ToUpper(ss))
	fmt.Println(strings.ToLower(ss))
	// Funciones de busqueda de cadena
	/* strings.HasPrefix 	Realiza búsquedas en la cadena desde el principio.
	strings.HasSuffix 	Realiza búsquedas en la cadena desde el final.
	strings.Contains 	Realiza búsquedas en cualquier parte de la cadena.
	strings.Count 	Cuenta la cantidad de veces que aparece la cadena.*/
	fmt.Println(strings.HasPrefix(ss, "Sammy"))
	fmt.Println(strings.HasSuffix(ss, "Shark"))
	fmt.Println(strings.Contains(ss, "Sh"))
	fmt.Println(strings.Count(ss, "S"))
	// Determina la longuitud de la cadena
	openSource := "Jhon contributes to open source."
	fmt.Println(len(openSource))
	// manipulación de Cadenas, strings.Join, strings.Split, strings.ReplaceAll
	fmt.Println(strings.Join([]string{"sharks", "crustaceans", "plankton"}, ","))
	balloon := "Sammy has a balloon."
	s := strings.Split(balloon, " ")
	fmt.Printf("%q", s)
	fmt.Println()
	// strings Filed ignora espacios en blanco
	data := "  username password   email	date"
	fields := strings.Fields(data)
	fmt.Printf("%q", fields)
	// ReplaceAll toma uan cadena original y realiza una sustitución
	fmt.Println(strings.ReplaceAll(balloon, "has", "had"))
	// Eliminar espacios en blanco
	fmt.Println(strings.TrimSpace(data)) // que pasa aqui ?
	fmt.Println(strings.TrimSpace(" \t\n Hello, Gophers \n\t\r\n"))
	// Trim retorna un slice del string quitando
	fmt.Println(strings.Trim("¡¡¡Hello, Gophers!!!", "!¡"))
	// Builder para construir un string
	words := []string{"Hola", " ", "Mundo"}

	var builder strings.Builder

	for _, w := range words {
		builder.WriteString(w)
	}

	result := builder.String()
	fmt.Println(result)
	// Comparar tamaño

	// create three strings
	string1 := "Programiz"
	string2 := "Programiz Pro"
	string3 := "Programiz"

	// compare strings
	fmt.Println(string1+" vs "+string2+":", strings.Compare(string1, string2)) // -1
	fmt.Println(string2+" vs "+string3+":", strings.Compare(string2, string3)) // 1
	fmt.Println(string1+" vs "+string3+":", strings.Compare(string1, string3)) // 0
}

// Iterar sobre un String
func TestStringIterationWithRange() {
	fmt.Println()
	namestring := "omar barra"
	for _, v := range namestring {
		fmt.Printf("%q\n", v)
	}
}

// Compara igualdad sin importar case
func TestStringEqualsCaseInsensitive() {
	got := "Go is Awesome!"
	expect := "go is awesome!"

	if !strings.EqualFold(got, expect) {
		fmt.Printf("expected %s got %s", expect, got)
	}
}

func esPalindromo(s string) bool {
	s = strings.ToLower(s)
	izquierda := 0
	derecha := len(s) - 1

	for izquierda < derecha {
		// Avanzar izquierda si no es letra o número
		for izquierda < derecha && !unicode.IsLetter(rune(s[izquierda])) && !unicode.IsNumber(rune(s[izquierda])) {
			izquierda++
		}
		// Retroceder derecha si no es letra o número
		for izquierda < derecha && !unicode.IsLetter(rune(s[derecha])) && !unicode.IsNumber(rune(s[derecha])) {
			derecha--
		}

		if s[izquierda] != s[derecha] {
			return false
		}

		izquierda++
		derecha--
	}

	return true
}

func esAnagrama(palabra1, palabra2 string) bool {
	// contienen el mismo número de ocurrencias de cada letra en orden diferente
	if len(palabra1) != len(palabra2) {
		return false
	}
	cuenta := make(map[rune]int)

	for _, letra := range strings.ToLower(palabra1) {
		if unicode.IsLetter(letra) {
			cuenta[letra]++
		}
	}

	for _, letra := range strings.ToLower(palabra2) {
		if unicode.IsLetter(letra) {
			cuenta[letra]--
			if cuenta[letra] < 0 {
				return false
			}
		}
	}

	for _, valor := range cuenta {
		if valor != 0 {
			return false
		}
	}

	return true
}
func esIsograma(palabra string) bool {
	// No tiene letras repetidas
	usado := make(map[rune]bool)
	for _, letra := range strings.ToLower(palabra) {
		if letra == '-' || letra == ' ' {
			continue
		}
		if _, ok := usado[letra]; ok {
			return false
		}
		usado[letra] = true
	}
	return true
}
func InvertirString(s string) string {
	// Convierte el string a un slice de runas para manejar adecuadamente caracteres Unicode.
	runas := []rune(s)
	for i, j := 0, len(runas)-1; i < j; i, j = i+1, j-1 {
		// Intercambia los elementos.
		runas[i], runas[j] = runas[j], runas[i]
	}
	// Convierte el slice de runas de vuelta a string.
	return string(runas)
}
